fix: child span not acquiring transaction lock in some cases

kruskall opened the pull request:
Spans from the same transaction share the dropped spans map. If they end concurrency a race condition could happen when updating the map.
Add a RW lock to prevent that.

Closes https://github.com/elastic/apm-agent-go/issues/1484

marclop reviewed (commented):
Would you mind adding a test covering the bug?

kruskall commented:
I've added a test and discovered more issues. It seems this is affecting more than just the dropped spans.

I've pushed some changes and added a test.

marclop commented on span.go:
```diff
@@ -397,6 +397,10 @@ func (s *Span) End() {
 			s.dropWhen(true)
+			if s.tx != nil {
+				s.tx.mu.Lock()
+				defer s.tx.mu.Unlock()
```
Reading the godoc in `span.end()`, `s.tx.transactionData.mu` also needs to be held:

https://github.com/elastic/apm-agent-go/blob/326e364d6e52ea9601deb41a0004b633e1a03051/span.go#L466-L471

```suggestion
				defer s.tx.mu.Unlock()
				if !s.tx.ended() {
					s.tx.TransactionData.mu.Lock()
					defer s.tx.TransactionData.mu.Unlock()
				}
```

marclop commented on span_test.go:
```diff
@@ -1044,6 +1044,40 @@ func TestCompressSpanSameKindConcurrent(t *testing.T) {
+
+	// Wait until all the spans have ended.
+	wg.Wait()
+
```
The parent and transaction should also end:

```suggestion
	parent.End()
	tx.End()
```

marclop commented on span_test.go:
```diff
@@ -1044,6 +1044,40 @@ func TestCompressSpanSameKindConcurrent(t *testing.T) {
+func TestDroppedSpanParentConcurrent(t *testing.T) {
+	// This test verifies there aren't any deadlocks on calling
+	// span.End(), Parent.End() and tx.End().
+	tracer := apmtest.NewRecordingTracer()
```
```suggestion
	tracer := apmtest.NewRecordingTracer()
	defer tracer.Close()
```

marclop reviewed (commented):
Thanks, this looks better!

simitt reviewed (commented):
Once this is ready to merge, please also add a changelog entry

kruskall commented:
@simitt could you clarify the above question ?

axw commented:
@kruskall in the past we have added entries to the Unreleased section. We've not been consistent about it though, and have sometimes just added the entries retrospectively at release time. Let's merge this and create a patch release - either adding the entry now, or creating the section at release time.

simitt commented:
Apologies @kruskall , I missed your message.
